numpy/_utils/__init__.py

```diff
@@ -26,6 +27,12 @@ def example():
     """
     def decorator(func):
         if module is not None:
+            if isinstance(func, type):
+                try:
+                    func._module_file = sys.modules.get(func.__module__).__file__
```

Comment:
Total nitpick, but just `func._module_file = sys.modules[func.__module__].__file__` would be fine since you check for `KeyError`.